### ISSUE TITLE
[Feature] Verifying Yarn release binaries

### ISSUE THREAD

**Issue by jonahsnider** [yarnpkg/berry#3636]
- [ ] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [x] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**

Without a way to verify release files Yarn users may be vulnerable to PRs which introduce malicious code to the local Yarn binary.

Thankfully, this is not a problem unique to Yarn. In the past this has been solved with release signing (not really possible for Yarn) or by publishing the hashes of release files. [Gradle wrappers are susceptible to the same issue](https://blog.gradle.org/gradle-wrapper-checksum-verification-github-action).

**Describe the solution you'd like**

The solution Gradle found was to publish the hashes of release binaries and [create CI integrations to help users verify them automatically](https://github.com/marketplace/actions/gradle-wrapper-validation).

Yarn should publish hashes alongside releases and create tooling for verifying them. If possible something similar should be done for plugins.

**Describe the drawbacks of your solution**

Including hashes with each release introduces some extra complexity to the release process. It would also require maintaining verification tooling/CI integrations.

**Comment by peey** [comment]
@arcanis is it possible to extract yarn version and use that version to get a trusted release binary from a trusted server (e.g. wherever yarn downloads from in the first place), without invoking the yarn binary in package?

If someone wishes to implement any kind of solution to this issue (matching a hash / downloading independent release binary) inside of yarn or as an independent tool (even at corepack level as you suggested), I believe that's the absolute minimum support that'll be required from within yarn itself.

**Comment by arcanis** [comment]
> I can verify npm, yarn (1.x), and pnpm tarballs against checksums and signatures published by the registry (e.g. npm view yarn dist).

If we assume that the https layer between https://repos.yarnpkg.com and you can be broken and a MITM attack performed, can't we say the same from the layer between https://registry.npmjs.org and you? Since the registry-provided checksums come from the same domain that serves the tarballs, I'm not sure they provide a significantly better security 🤔

(I'm however not an expert, so if I'm very wrong I'm interested to learn about anything I might be missing)

> the yarn 2+ binary isn't published as a part of any npm package.

It's published as part of [`@yarnpkg/cli-dist`](https://yarnpkg.com/package/@yarnpkg/cli-dist?files).

**Comment by joshwlewis** [comment]
> If we assume that the https layer between https://repos.yarnpkg.com/ and you can be broken and a MITM attack performed, can't we say the same from the layer between https://registry.npmjs.org/ and you? Since the registry-provided checksums come from the same domain that serves the tarballs, I'm not sure they provide a significantly better security - an attacker could just MITM both the checksum and the tarball 🤔

Yes, you are correct about the checksum portion. However, the registry's public signing key (available [here](https://registry.npmjs.org/-/npm/v1/keys)) can be stored in code or config at a time when uncomprimised. Later, when the connection might be comprimised, verifying the signature (from `npm view @yarnpkg/cli-dist dist.signatures`) should fail, because the attacker can't generate a valid signature without the private signing key.

> It's published as part of [@yarnpkg/cli-dist](https://yarnpkg.com/package/@yarnpkg/cli-dist?files).

Perfect, this should work for me! I searched several times for a package that might work, but I think npmjs.com was truncating the search results.

**Comment by MatthiasKunnen** [comment]
Perhaps this should be extended to the plugins downloaded to `.yarn/plugins` such as `@yarnpkg/plugin-workspace-tools`?

**Comment by hichemfantar** [comment]
is there no built in way for yarn to verify binary integrity?
